### ISSUE TITLE
Module Management

### ISSUE THREAD

**Issue by kbserm** [OAGi/Score#801]

**Comment by hakjuoh** [comment]
@kbserm @rekuku @asdimitriadis 

![module_2 0](https://user-images.githubusercontent.com/16102116/85943113-dcdd1580-b8fb-11ea-941c-f122790813e4.png)

The following items are newly added tables.

`module_dir`: This contains directories for the module. It is defined in a self-contained manner (i.e. hierarchy structure). Let say we define a module with '/Model/Nouns/AcknowledgeBOM' path. Then, there are three `module_dir` records for '/' (root), 'Model', and 'Nouns'.

`module_set`: This contains module sets. Initially, there is a record for 'OAGIS 10.6 Canonical XML Schema'.
`module_set_assignment`: This is a relationship table to make a connect between `module` and `module_set`.
`module_set_release`: This table depicts a relationship between `release` and `module_set` as a release could have multiple module sets.
`module_cc_manifest`: There are five tables for each cc release manifest table; `module_acc_manifest`, `module_asccp_manifest`, `module_bccp_manifest`, `module_dt_manifest`, and `module_code_list_manifest`. Note that `module_ascc_manifest`, `module_bcc_manifest`, `module_dt_sc_manifest`, and `module_code_list_value_manifest` don't need to be existed since it must share the module with the parent component.

Two existing tables for the module management also have some changes.
`module`: `module_dir_id` column is added and it cannot be null. A path of the module can express as `concat(module_dir.path, '/', module.name)`.
`module_dep`: It replaces `depending_module_id` and `depended_module_id` with `depending_module_set_assignment_id` and `depended_module_set_assignment_id`, respectively.

**Comment by kbserm** [comment]
@hakjuoh I guess you need module agency_id_list_manifest as well.

**Comment by kbserm** [comment]
If we are not making UI in 2.0, we need to test the database schema that it can hold modules information for at least 2 OAGIS releases.

We also might want to update the OAGIS model schema generation code so that it can generate/express OAGIS releases.

**Comment by kbserm** [comment]
There is a motivation to do this b/c if we can generate schema, it can be used to validate the release.

**Comment by kbserm** [comment]
Also, developer user should be able to generate schema even in the release is in release draft state. Need to go back and analyze whether the data model supports this.

**Comment by rekuku** [comment]
Include Scheme generation

**Comment by kbserm** [comment]
On the "Edit Module Set Release" page, change the button labels. "Export" -> "Express CCs XML Schema", "Assign CCs" -> "CCs Assignment"

**Comment by kbserm** [comment]
I logged in as an end user and still see the Module menu. The ability to create/edit module content has been disabled, but there is an error when clicking to open a Module Set. On of the patterns, we agreed previously is that in general we should let user who has no right to a function to be able to see the menu but that the menu is disabled. I think we should do the same here. There seems to be no point in having the end user to be able to see the module management information but cannot do anything.

So in summary at this point, I would say either hide or block access to the Module functions from end user or have the Module menu disabled. I think we should do the former b/c seeing the Module menu doesn't really inform the end user much.

**Comment by kbserm** [comment]
I reopened this because I want to move it to another release. I think there are a little more to straighten out, but I didn't want this to block 2.2 release.

**Comment by kbserm** [comment]
![image](https://user-images.githubusercontent.com/7863166/158508839-def066f4-fb11-4f2b-9886-13b37f64a3ee.png)


In this dialog, please the labels as follows:
Add module/directory to "Root" => Add a file or directory to "Root"
Copy from exist module set => Copy from a module set

**Comment by kbserm** [comment]
![image](https://user-images.githubusercontent.com/7863166/158567878-6e3c9064-db46-4dc7-b630-7194a8e6a0d3.png)

Can the Copy button be next to the Cancel button? On my screen, I have to scroll down to see the Copy button.

Also, every time a copying is made the whole thing collapse back to root, kinda user-unfriendly.

**Comment by kbserm** [comment]
@asdimitriadis User Guide says that namespace of a module file is optional. It is optional when it is first created but then Score assign a nmsp after saving the module file. Upon editing it again the nmsp cannot be deleted. So I think it is required per the design. Now I'm not sure which namespace the system would automatically assign if there are multiple std nmsp. Could u iron this out.

Also the user guide said Version is required but it is option. I think per the design it is option.

**Comment by kbserm** [comment]
![image](https://user-images.githubusercontent.com/7863166/158582346-8f3135b0-2fd4-4181-9901-b174f7cf3c34.png)

The name of this page should be changed to "Core Component Assignment"

**Comment by kbserm** [comment]
The UI on the CC Assignment section does not work well. First, right now the top part (red mark in the screenshot below) where the filters are also scroll, it shouldn't. Seconding, I cannot use the Ctrl key to check multiple entities. This make the section harder. It looks like the UI components used in the Release Management page might be better. We don't have to change this in 2.3. You can refactor this as an issue in a later release.

![image](https://user-images.githubusercontent.com/7863166/158601334-2e3e1a17-87f1-4761-9ff3-5de5dfa2ed2b.png)

**Comment by asdimitriadis** [comment]
I am closing this issue since we do not have any other issue for the module management. 
All issues related to the module management are addressed.
